server/room.ts

```diff
@@ -39,214 +46,115 @@ export class Room {
-      });
-      socket.on(
-        'signalSS',
-        (data: { to: string; sharer: boolean; msg: string }) => {
-          io.of(roomId).to(data.to).emit('signalSS', {
-            from: socket.id,
-            sharer: data.sharer,
-            msg: data.msg,
-          });
-        }
-      );
+  sendChatMessage = (socket: Socket, message: string) => {
+    if ((message && message.length > 65536) || !socket) {
+      // TODO add some validation on client side too so we don't just drop long messages
+      return;
+    }
+    if (process.env.NODE_ENV === 'development' && message === '/clear') {
+      this.chat.length = 0;
+      this.io.of(this.roomId).emit('chatinit', this.chat);
+      return;
+    }
+    const chatMsg = { id: socket.id, msg: message };
+    this.addChatMessage(socket, chatMsg);
+  };
 
-      socket.on('disconnect', () => {
-        let index = this.roster.findIndex((user) => user.id === socket.id);
-        const removed = this.roster.splice(index, 1)[0];
-        io.of(roomId).emit('roster', this.roster);
-        if (removed.isScreenShare) {
-          // Reset the room state since we lost the screen sharer
-          this.cmdHost(socket, '');
-        }
-        delete this.tsMap[socket.id];
-        // delete nameMap[socket.id];
-      });
+  startVBrowser = async (socket: Socket) => {
+    if (this.vBrowser || !socket) {
+      // Maybe terminate the existing instance and spawn a new one
+      return;
+    }
+    redisCount('vBrowserStarts');
+    this.cmdHost(socket, 'vbrowser://');
+    const assignment = await this.vmManager.assignVM();
+    if (!assignment) {
+      this.cmdHost(socket, '');
+      this.vBrowser = undefined;
+      return;
+    }
+    this.vBrowser = assignment;
+    this.roster.forEach((user, i) => {
+      if (user.id === socket.id) {
+        this.roster[i].isController = true;
+      } else {
+        this.roster[i].isController = false;
+      }
     });
+    this.cmdHost(
+      socket,
+      'vbrowser://' + this.vBrowser.pass + '@' + this.vBrowser.host
+    );
+    this.io.of(this.roomId).emit('roster', this.roster);
+  };
+
+  async stopVBrowser(socket: Socket) {
+    const assignTime = this.vBrowser && this.vBrowser.assignTime;
+    const id = this.vBrowser && this.vBrowser.id;
+    this.vBrowser = undefined;
+    this.roster.forEach((_USER, i) => {
+      this.roster[i].isController = false;
+    });
+    if (this.videoPlaylist.length > 0) {
+      this.nextVideo();
+    } else {
+      this.cmdHost(socket, '');
+      this.isRoomDirty = true;
+    }
+    if (redis && assignTime) {
+      await redis.lpush('vBrowserSessionMS', Number(new Date()) - assignTime);
+      await redis.ltrim('vBrowserSessionMS', 0, 24);
+    }
+    if (id) {
+      try {
+        await this.vmManager.resetVM(id);
+      } catch (e) {
+        console.error(e);
+      }
+    }
   }
 
   serialize() {
     return JSON.stringify({
       video: this.video,
+      videoDuration: this.videoDuration,
```

Comment:
are we setting this for all video types or only youtube videos? I assume for URLs we'd have to fetch the metadata from the video file

Reply:
Right now only youtube videos, yeah. Do you know a good way to get a videos duration from just the HTTP link?

Reply:
I don't think so, without downloading some of the video's file data.